### PR TITLE
Experiment Versioning | Code updates

### PR DESCRIPTION
**Pull request by SmittieC**
This PR is the code updates part of [the orignal PR](https://github.com/dimagi/open-chat-studio/pull/655). That PR is too big and confusing, so this is the new one that should be reviewed. I'm keeping it in draft until we have some UI changes to go along with it.

I will update this PR once the [UI changes](https://github.com/dimagi/open-chat-studio/pull/642) is in.

## Description

This is the code changes to use either the working or versioned experiments. This branch should be tested rigorously.

## User Impact

There should be no impact if there are no experiment versions.

### Demo

Stay tuned...

### Docs

Stay tuned...

### TODO
- More tests!
- Experiment deletion
- Related model deletion (should notify the user and deprecate versions that uses that)
- API to specify which version to use
- Web user to specify which version to use
- Events
- Triggers
- Scheduled Messages

### CODE REVIEW

File: apps/experiments/views/source_material.py
Comment (stephherbers) [inline apps/experiments/views/source_material.py:36]: so this only show the source materials associated with a released version? I feel like maybe we should exclude material associated with an archived experiment, but why the working?

File: apps/experiments/views/safety.py
Comment (stephherbers) [inline apps/experiments/views/safety.py:33]: some comment here: maybe I'm missing the point of this logic rather than if the exp is archived

**Comment by stephherbers** [comment]
good start-- noting too that when filtering for versions we should ensure that `is_archived` is not false for either the working or the version being queried-- maybe useful to modify the helper functions

BUT since this FFed and we actually can't archive experiments yet, we can hold off to a separate PR when we add in that functionality but I'll leave it up to you

File: apps/experiments/decorators.py
Comment (snopoke) [inline apps/experiments/decorators.py:29]: fine to leave this call here but just confirming that this should already be the working version right?

File: apps/experiments/models.py
Comment (snopoke) [inline apps/experiments/models.py:692]:
```suggestion
        - they are not children of the current experiment
```

File: apps/experiments/models.py
Comment (snopoke) [inline apps/experiments/models.py:704]: I think this would be more efficient as `working_version_id=parent.id`

File: apps/experiments/views/experiment.py
Comment (snopoke) [inline apps/experiments/views/experiment.py:888]: won't this always be the working version?

File: apps/experiments/views/experiment.py
Comment (snopoke) [inline apps/experiments/views/experiment.py:977]: I think it will always be the default. The only time you can use a non-default is if you're an experiment admin in which case you wouldn't use an invite.

File: apps/experiments/views/experiment.py
Comment (snopoke) [inline apps/experiments/views/experiment.py:980]: Is this handled in the channel?

**Review by snopoke — COMMENTED** [review]
I can see it's going to be confusing to know when we're dealing with a version vs the default. I wonder if there's a clearer way to differentiate in code where we expect the working version vs a version vs either.

File: apps/experiments/decorators.py
Comment (SmittieC) [reply apps/experiments/decorators.py:29]: Not necessarily in the future, but for now yes. If we have a button that allows the user to chat to a specific version, then this might be the versioned experiment

File: apps/experiments/models.py
Comment (SmittieC) [reply apps/experiments/models.py:704]: Ah yes, good catch thanks

File: apps/experiments/views/experiment.py
Comment (SmittieC) [reply apps/experiments/views/experiment.py:980]: Correct. When we start the session (this happens when we invite the user), the seed message is being generated, so doing it here essentially double generated the seed message

File: apps/experiments/views/experiment.py
Comment (SmittieC) [reply apps/experiments/views/experiment.py]: Correct. I updated the views to always assume the `experiment_id` is the working experiment (https://github.com/dimagi/open-chat-studio/pull/671/commits/ebaac498dec2fe45ee72c80f7782421a3c55b2a1)

File: apps/experiments/views/experiment.py
Comment (snopoke) [inline apps/experiments/views/experiment.py:996]: This doesn't seem to be used by the form. It's unclear to me if it's needed at all and if so whether it should be the working or default version

File: apps/slack/slack_listeners.py
Comment (snopoke) [inline apps/slack/slack_listeners.py:68]: I think this should be the working version

File: apps/channels/views.py
Comment (SmittieC) [reply apps/channels/views.py:101]: Correct. Yes forgot to update this piece. I am still to update the API endpoints to support multiple versions as well

File: apps/experiments/views/experiment.py
Comment (SmittieC) [reply apps/experiments/views/experiment.py:806]: Nope, not true. Another one I didn't update. Thanks for your hawk eyes

File: apps/experiments/models.py
Comment (SmittieC) [reply apps/experiments/models.py]: I think I'm rather going to take this out, since [this PR](https://github.com/dimagi/open-chat-studio/pull/676) is in the pipeline. Archiving a single version is supported by that PR, and I suppose it wouldn't make much sense to have an archive button for the working experiment. If we delete it, everything related goes, which is fine.

File: apps/experiments/models.py
Comment (SmittieC) [reply apps/experiments/models.py]: This is redundant now. See [this comment](https://github.com/dimagi/open-chat-studio/pull/671#discussion_r1754815245)

File: apps/experiments/models.py
Comment (SmittieC) [reply apps/experiments/models.py]: https://github.com/dimagi/open-chat-studio/pull/671/commits/a46bb96e3050ca97ae118beaa030654f0d883a96

File: apps/experiments/views/experiment.py
Comment (SmittieC) [reply apps/experiments/views/experiment.py]: https://github.com/dimagi/open-chat-studio/pull/671/commits/a46bb96e3050ca97ae118beaa030654f0d883a96

File: apps/experiments/views/experiment.py
Comment (SmittieC) [reply apps/experiments/views/experiment.py]: You're right, it's not being used. Ripping it out...

File: apps/slack/slack_listeners.py
Comment (SmittieC) [reply apps/slack/slack_listeners.py:68]: See https://github.com/dimagi/open-chat-studio/pull/671#discussion_r1754850320

File: apps/experiments/views/experiment.py
Comment (snopoke) [reply apps/experiments/views/experiment.py:698]: I agree with that except that it's not here where you make that call, it's where you pass the experiment to the task (which is a few lines below). The session should always be linked to the working version.

File: apps/slack/slack_listeners.py
Comment (snopoke) [reply apps/slack/slack_listeners.py:68]: Same as reply to other comment

**Comment by stephherbers** [comment]
still going through the review but wanted to add that the test plan sounds good and the scope of the PR

File: apps/experiments/views/experiment.py
Comment (SmittieC) [reply apps/experiments/views/experiment.py]: https://github.com/dimagi/open-chat-studio/pull/671/commits/40f779b2f3819d6fa42a7426ab0177d2530e9611

**Comment by SmittieC** [comment]
I built a bot on an unversioned experiment and went through basically all things affected by these changes. Then I made a version and did the same. No breakages (or at least those that did break were fixed). I think this is safe to merge as-is

File: apps/slack/slack_listeners.py
Comment (SmittieC) [reply apps/slack/slack_listeners.py]: https://github.com/dimagi/open-chat-studio/pull/671/commits/40f779b2f3819d6fa42a7426ab0177d2530e9611

File: apps/experiments/models.py
Comment (snopoke) [inline apps/experiments/models.py:1134]: nit:
```suggestion
        return self.experiment.default_version
```

**Review by snopoke — DISMISSED** [review]
Nothing blocking

File: apps/chat/channels.py
Comment (SmittieC) [reply apps/chat/channels.py]: https://github.com/dimagi/open-chat-studio/pull/671/commits/0a38ec682a07396da7c21727fd5e13b6d0f968ef

**Pull request merged**
